scripts/assemble_changelog.py

```diff
@@ -0,0 +1,284 @@
+#
+# This file is part of Mbed Crypto (https://tls.mbed.org)
+
+import argparse
+import glob
+import os
+import re
+import sys
+
+class InputFormatError(Exception):
+    def __init__(self, filename, line_number, message, *args, **kwargs):
+        message = '{}:{}: {}'.format(filename, line_number,
+                                     message.format(*args, **kwargs))
+        super().__init__(message)
+
+class LostContent(Exception):
+    def __init__(self, filename, line):
+        message = ('Lost content from {}: "{}"'.format(filename, line))
+        super().__init__(message)
+
+STANDARD_SECTIONS = (
+    b'Interface changes',
+    b'Default behavior changes',
+    b'Requirement changes',
+    b'New deprecations',
+    b'Removals',
+    b'New features',
+    b'Security',
+    b'Bug fixes',
+    b'Performance improvements',
+    b'Other changes',
+)
+
+class ChangeLog:
+    """An Mbed Crypto changelog.
+
+    A changelog is a file in Markdown format. Each level 2 section title
+    starts a version, and versions are sorted in reverse chronological
+    order. Lines with a level 2 section title must start with '##'.
+
+    Within a version, there are multiple sections, each devoted to a kind
+    of change: bug fix, feature request, etc. Section titles should match
+    entries in STANDARD_SECTIONS exactly.
+
+    Within each section, each separate change should be on a line starting
+    with a '*' bullet. There may be blank lines surrounding titles, but
+    there should not be any blank line inside a section.
+    """
+
+    _title_re = re.compile(br'#*')
+    def title_level(self, line):
+        """Determine whether the line is a title.
+
+        Return (level, content) where level is the Markdown section level
+        (1 for '#', 2 for '##', etc.) and content is the section title
+        without leading or trailing whitespace. For a non-title line,
+        the level is 0.
+        """
+        level = re.match(self._title_re, line).end()
+        return level, line[level:].strip()
+
+    def add_sections(self, *sections):
+        """Add the specified section titles to the list of known sections.
+
+        Sections will be printed back out in the order they were added.
+        """
+        for section in sections:
+            if section not in self.section_content:
+                self.section_list.append(section)
+                self.section_content[section] = []
+
+    def __init__(self, input_stream):
+        """Create a changelog object.
+
+        Populate the changelog object from the content of the file
+        input_stream. This is typically a file opened for reading, but
+        can be any generator returning the lines to read.
+        """
+        self.header = []
+        self.section_list = []
+        self.section_content = {}
+        self.add_sections(*STANDARD_SECTIONS)
+        self.trailer = []
+        self.read_main_file(input_stream)
+
+    def read_main_file(self, input_stream):
```

Comment:
I added a comment. Creating a new section for a new version is outside the scope of this function and I intended it to be in a follow-up pull request which, indeed, adds a command-line option that takes the version number as an argument. This script can be used at any time, not just to make a release.

Reply:
I added some logic to create a level-2 section if it seems to be needed. This way, you can run the script with no argument at any time, and it will populate the “Unreleased changes” section, creating it if it doesn't exist. At release time, first run the script, then edit the title “Unreleased changes” to “Mbed Wossname 3.14”. I think it makes more sense this way, but I'm willing to be talked into a different workflow.

Reply:
Thanks for adding the comments, I think that's much clearer now.

Regarding the logic, I think what you implemented might be more flexible than what we need (I don't see ourselves running that script too often between releases), but this flexibility comes as no cost in terms of code complexity or usage complexity, and might turn out to be needed contrary to my expectations, so I'm happy with it.